feat: linear gradient `px` and transition hint syntax support

intergalacticspacehighway opened the pull request:
<!-- Thanks for submitting a pull request! We appreciate you spending the time to work on these changes. Please provide enough information so that others can review your pull request. The three fields below are mandatory. -->

## Summary:
- Adds support for color transition hint syntax in linear gradients. e.g. `linear-gradient(red, 20%, green)`
- Adds `px` support. Combination of `px` and `%` also works.
- Simplified color stops parsing.
- The `processColorTransitionHint` and `getFixedColorStops` is moved to native code so it can support combination of `px` and `%` units as it requires gradient line length, which is derived from view dimensions and gradient line angle.
- Follows CSS [spec](https://drafts.csswg.org/css-images-4/#coloring-gradient-line) (Refer transition hint section) and implementation is referred from [blink engine source](https://github.com/chromium/chromium/blob/a296b1bad6dc1ed9d751b7528f7ca2134227b828/third_party/blink/renderer/core/css/css_gradient_value.cc#L240).


<!-- Explain the **motivation** for making this change. What existing problem does the pull request solve? -->

## Changelog:

[GENERAL] [ADDED] - Linear gradient color transition hint syntax and `px` unit support.

## Test Plan:

Added testcase in processBackgroundImage-test.ts and example in LinearGradientExample.js
<!-- Demonstrate the code is solid. Example: The exact commands you ran and their output, screenshots / videos if the pull request changes the user interface. -->


<img width="500" alt="Screenshot 2025-01-05 at 11 38 13 PM" src="https://github.com/user-attachments/assets/62858bb7-1dbf-40cf-8dd4-ec0daf84ac1b" />


## Todo

Add testcases for `getFixedColorStops` and `processColorTransitionHint` in native code for both platforms. That's the only downside of moving it out of JS 🤦

joevilches commented on packages/react-native/Libraries/StyleSheet/processBackgroundImage.js:
```diff
@@ -49,33 +51,52 @@ export default function processBackgroundImage(
-              // If a position is invalid, return an empty array and do not apply gradient. Same as web.
-              return [];
-            }
+        const positions = colorStop.positions;
+        // Color transition hint syntax (red, 20%, blue)
+        if (
+          colorStop.color == null &&
+          Array.isArray(positions) &&
+          positions.length === 1
+        ) {
+          const position = positions[0];
+          if (typeof position === 'string' && position.endsWith('%')) {
+            processedColorStops.push({
+              color: null,
+              position: parseFloat(position) / 100,
+            });
+          } else {
+            // If a position is invalid, return an empty array and do not apply gradient. Same as web.
+            return [];
```
I think we can change this to share some code better. Do transition hints have to be percentages? Regardless we are doing a lot of the same things, can you try to get rid of the top level if/else here and share some of this logic?

intergalacticspacehighway replied:
> Do transition hints have to be percentages?

They are actually `length-percentage` in the [spec](https://developer.mozilla.org/en-US/docs/Web/CSS/length-percentage). But i only added `%` support for now in color stops. Linear gradient support mixing `%` and `px` units which require us to know the dimension, for this we'll have to move color stop parsing to native side. I wanted to avoid that change until the new css parser. 

>  can you try to get rid of the top level if/else here and share some of this logic?

yeah i kinda prefer the explicit checks, also each branch has a tiny comment on top, but i will take another look!

joevilches commented on packages/react-native/Libraries/StyleSheet/processBackgroundImage.js:
```diff
@@ -22,11 +22,13 @@ type LinearGradientDirection =
   direction: LinearGradientDirection,
   colorStops: $ReadOnlyArray<{
-    color: ProcessedColorValue,
+    color: ColorStopColor,
```
I would comment in what scenarios we can have a null color, or otherwise change the typing to make that super clear like 
```
$ReadOnlyArray<{
  colorStop | transitionHint
}>
```
I think they are similar enough to just roll with the comment but it should be clear that that syntax is supported

intergalacticspacehighway replied:
Added a [comment](https://github.com/facebook/react-native/pull/48410/commits/6913cdb54e02781ac357bcb64170331344ef1ebc#diff-645b1d95fc4a2f92f0a8a2e695cd9b4c31ebc977adac0a38b39c46946f3fd530R25). Thanks!

joevilches commented on packages/react-native/ReactCommon/react/renderer/components/view/conversions.h:
```diff
@@ -1331,7 +1331,13 @@ inline void fromRawValue(
               ColorStop colorStop;
               colorStop.position = (Float)(positionIt->second);
-              fromRawValue(context, colorIt->second, colorStop.color);
+              if (colorIt->second.hasValue()) {
```
what is the default value of the color in ColorStop? Will we be able to detect that this is a transition hint vs something like the color black?

intergalacticspacehighway replied:
here if JS color is `null` that only happens if user passed transition hint. Then the color in ColorStop gets default SharedColor value i.e. `HostPlatformColor::UndefinedColor`. So for black color it should be alright!

joevilches commented on packages/react-native/React/Fabric/Utils/RCTLinearGradient.mm:
```diff
@@ -131,4 +133,132 @@ static CGFloat getAngleForKeyword(GradientKeyword keyword, CGSize size)
+    auto &colorStop = colorStops[i];
+    // Skip if not a color hint
+    if (colorStop.color) {
+      continue;
```
again, curious if a color like black breaks this? Regardless I feel like it would be more future proof and more obvious if you just changed the color in ColorStop to be a std::optional or such

intergalacticspacehighway replied:
color in ColorStop is SharedColor instance which defaults to `HostPlatformColor::UndefinedColor` and it has [bool overload](https://github.com/facebook/react-native/blob/7e029b0dcf6d1a6455a8a6343457b70e353d0ff6/packages/react-native/ReactCommon/react/renderer/graphics/Color.h#L48) to check for undefined color. So i think black or such colors should be fine?

joevilches commented on packages/react-native/Libraries/StyleSheet/__tests__/processBackgroundImage-test.js:
```diff
@@ -697,4 +697,34 @@ describe('processBackgroundImage', () => {
+      {color: null, position: 0.4},
+      {color: processColor('blue'), position: 1},
+    ]);
+  });
```
can you add some more? Hints that are invalid (> 100%, non percents, multiple in a row, etc). Some longer ones with more stops and more hints just to stress test things a bit, etc. The code has a lot of branching logic due to the complexity of the syntax we are allowing, so I think we should err on the side of excessive testing to raise our confidence level here

intergalacticspacehighway replied:
Agree, [added](https://github.com/facebook/react-native/pull/48410/commits/6913cdb54e02781ac357bcb64170331344ef1ebc).

intergalacticspacehighway commented:
@joevilches i added the `px` support as well and updated the PR description. sorry it increased the review surface and changes. But it was needed for web parity.

joevilches commented on packages/react-native/ReactAndroid/src/main/java/com/facebook/react/uimanager/style/LinearGradient.kt:
```diff
@@ -7,16 +7,42 @@
 import kotlin.math.tan
+import android.graphics.LinearGradient as AndroidLinearGradient
+
+private enum class UnitType {
+  Point,
+  Percent,
+  Undefined
+}
+
+private data class ValueUnit(
+  val value: Float = 0.0f,
+  val unit: UnitType = UnitType.Undefined
+)
+
+private data class ColorStop(
+  var color: Int? = null,
+  val position: ValueUnit
+)
```
@intergalacticspacehighway got some internal comments mentioning we should use LengthPercentage here for `position`'s type: https://github.com/facebook/react-native/blob/main/packages/react-native/ReactAndroid/src/main/java/com/facebook/react/uimanager/LengthPercentage.kt

intergalacticspacehighway replied:
Sure, [updated](https://github.com/facebook/react-native/pull/48410/commits/a1048a21ee728457e46f096fbec1a2bc5a9e88b7). I missed checking that out. Thanks!

jorge-cab commented on packages/react-native/ReactAndroid/src/main/java/com/facebook/react/uimanager/LengthPercentage.kt:
```diff
@@ -20,9 +20,12 @@ public enum class LengthPercentageType {
 
 public data class LengthPercentage(
-    private val value: Float,
+    private val _value: Float,
     public val type: LengthPercentageType,
 ) {
+  public val value: Float
+    get() = _value
```
I think we can just change value to be public. We don't have to worry about the setter since its a val

intergalacticspacehighway replied:
Cool. Updated. Thanks for reviewing.

jorge-cab commented on packages/react-native/ReactAndroid/src/main/java/com/facebook/react/uimanager/style/LinearGradient.kt:
```diff
@@ -220,23 +196,23 @@ internal class LinearGradient(
-          0 -> newPosition = ValueUnit(0f, UnitType.Point)
-          colorStops.size - 1 -> newPosition = ValueUnit(1f, UnitType.Point)
+          0 -> newPosition = 0f
+          colorStops.size - 1 -> newPosition = 1f
```
nit: 
```
newPosition = newPosition ?: when (i) {
    0 -> 0f
    colorStops.size - 1 -> 1f
}
```

joevilches commented on packages/react-native/ReactAndroid/src/main/java/com/facebook/react/uimanager/style/LinearGradient.kt:
```diff
@@ -134,4 +192,203 @@ internal class LinearGradient(
+    if (position == null) return null
+
+    return when (position.type) {
+      LengthPercentageType.POINT -> PixelUtil.toPixelFromDIP(position.value) / gradientLineLength
+      LengthPercentageType.PERCENT -> position.value / 100
```
are percents eval'ed relative to the gradientLineLength?

intergalacticspacehighway replied:
actually, `px` are relative to gradientLineLength 😅. android and iOS linear gradient APIs expect [0-1] value for color stop position. So we need [this](https://github.com/facebook/react-native/pull/48410/commits/5e31def8541048327a59e6bc65914c26768d748e#diff-39a947fba82999bc9e458030eb8395e411ea4a1564078e9af91174857f15e7edL388) kind of resolve . e.g. if user passes `50%` we just want to resolve it to `0.5` and If user passes `50px` then we need to resolve it relative to gradientLineLength `50 / gradientLineLength`

joevilches reviewed (changes requested):
A few more things with the new length percentage!

intergalacticspacehighway commented on packages/react-native/ReactAndroid/src/main/java/com/facebook/react/uimanager/LengthPercentage.kt:
```diff
@@ -20,7 +20,7 @@ public enum class LengthPercentageType {
 
 public data class LengthPercentage(
-    private val value: Float,
+    public val value: Float,
```
Agree. [Updated](https://github.com/facebook/react-native/pull/48410/commits/5e31def8541048327a59e6bc65914c26768d748e). 

> are percents eval'ed relative to the gradientLineLength?

actually, `px` are relative to gradientLineLength 😅. android and iOS linear gradient APIs expect [0-1] value for color stop position. So we need [this](https://github.com/facebook/react-native/pull/48410/commits/5e31def8541048327a59e6bc65914c26768d748e#diff-39a947fba82999bc9e458030eb8395e411ea4a1564078e9af91174857f15e7edL388) kind of resolve . e.g. if user passes `50%` we just want to resolve it to `0.5` and If user passes `50px` then we need to resolve it relative to gradientLineLength `50 / gradientLineLength`
